### DISCUSSION TITLE
How to get the argument names for data returned from call_method

### DISCUSSION THREAD

**Discussion by Rindustrial** [FreeOpcUa/opcua-asyncio#380] [Q&A] [answered]
I'm new to OPC - any help would be appreciated.

I am successfully getting data from calling a method like this:
VarValues = await obj.call_method("2:GetUserData", "E0001", "Pass123")
for x in VarValues:
    print(x)

Console Out:
000001
E0001
User1
Test engineer
['Manager', 'Testpool operator']
['Testpool configuration', 'Testpool calibration']

I'm sure this is an easy question - but how do I get the Names along with each matching output variable?

They are all listed for that Method in UaExpert as:
ns=2;s=MES.GetUserData.OutputArguments

**Comment by AndreasHeine** [comment, marked as answer]
just read the value of the "OutputArguments"

![Unbenannt](https://user-images.githubusercontent.com/56362817/103401206-40186080-4b48-11eb-9971-95629e27b872.PNG)

**Reply by Rindustrial** [reply]
Thank you Andreas!

This works, but I couldn't figure out how to receive the individual arguments as a two dimensional array so I could get each Name - so I did this ugly hack on the one dimensional argument array converted to strings so I could assemble an array of Names. Is there a better way?


    var = client.get_node("ns=2;s=MES.GetUserData.OutputArguments")
    VarNames = await var.read_value()
    Cnt = 0
    VarName = []
    for x in VarNames:      # get the names of each output argument
        #print(x)
        ThisValueSt = val_to_string(x)
        StEnd = ThisValueSt.find(",")
        VarName.append(ThisValueSt[14:StEnd])
        print(VarName[Cnt])
        Cnt = Cnt + 1

 


![image](https://user-images.githubusercontent.com/76796158/103421449-6bac4280-4b6a-11eb-83f2-c23202ed64ce.png)

**Reply by AndreasHeine** [reply]
this looks not right have you looked with debugger and made a breakpoint? can you make a screenshot and show it (from the debugger)?

this should be a list of ua.Argument()
https://github.com/FreeOpcUa/opcua-asyncio/blob/f189d87b2ab0ad1264fe6aaf05523d4fe27e3c6f/asyncua/ua/uaprotocol_auto.py#L3380

it should be possible to make something like:

```python
for arg in arglist:
        print(arg.Name) #<---
        print(arg.DataType)
        print(arg.ValueRank)
        print(arg.ArrayDimensions)
        print(arg.Description)
```

**Reply by AndreasHeine** [reply]
```python
    node = client.get_node("ns=2;i=4")
    args = await node.read_value()
    for arg in args: # <--- if you make a breakpoint here you could see the object
        print(arg.Name)
```

it should work!

![Unbenannt](https://user-images.githubusercontent.com/56362817/103422558-6834ad80-4ba2-11eb-8850-6a46a032df20.PNG)

**Reply by Rindustrial** [reply]
Ah - fantastic. Thanks!

I was trying unsuccessfully to treat it as an array like arg[0] and now I see that it is more like a dictionary arg.Name

Thank you very much!

You guys are doing fantastic work here.

Is there a way I can send a donation to the project?

**Reply by AndreasHeine** [reply]
> Is there a way I can send a donation to the project?

thats actually a good question ;) as far as i know not maybe in the future!? (@oroulet)

**you can fork and star this library!!!**
